### ISSUE TITLE
Update requirements for obsm embeddings

### ISSUE THREAD

**Issue by brianraymor** [chanzuckerberg/single-cell-curation#544]
# Context

See [**#single-cell-data-wrangling**](https://czi-sci.slack.com/archives/C024HCSH9PT/p1680135750661479?thread_ts=1680044347.557909&cid=C024HCSH9PT) and [**#single-cell-eng**](https://czi-sci.slack.com/archives/C023Q1APASK/p1680194226881869). 

During CXG conversion, validation for [embeddings](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#obsm-embeddings) **silently failed** allowing the ingestion pipeline to complete successfully. When the curator explored the dataset, an "Unexpected HTTP error" was shown because a 400 was returned by the internal [cellxgene REST API](https://github.com/chanzuckerberg/single-cell-explorer/blob/main/dev_docs/REST_API.md) due to missing embeddings in the CXG. 

Subsequent code review also revealed that both the `cellxgene-schema CLI `and the CXG conversion process validate embeddings with some duplication, but their validation requirements are different. 

1. [Embedding validation in CXG conversion](https://github.com/chanzuckerberg/single-cell-data-portal/blame/15da6106f6742704d3233d2ab6e128b231ce7ae8/backend/layers/processing/h5ad_data_file.py#L107)
2. [Embedding validation in `cellxgene-schema CLI `](https://github.com/chanzuckerberg/single-cell-curation/blob/79a78b41414441d6f79ecf57baa57e3cea0c5eb3/cellxgene_schema_cli/cellxgene_schema/validate.py#L997)

Please review and document the complete set of requirements in the schema.

**Comment by signechambers1** [comment]
@atarashansky do you have a proposal for this? You thought you would be able to complete by the end of the sprint which is today.

**Comment by atarashansky** [comment]
Validation requirements for embeddings:

 - OBSM key must start with `X_` and must have a suffix (`"X_{suffix}"`) that is at least one character long.
 - OBSM value (the embedding array) must be a numpy array.
 - The embedding array `dtype` must be a float, integer, or unsigned integer of any precision (8, 16, 32, or 64 bits).
 - The embedding array must have the same number of rows as the expression matrix and it must have at least 2 columns.
 - The embedding array must not have any "inf" values.
 - The embedding array must not contain all "nan" values.

**Comment by brianraymor** [comment]
@jahilton - can you review Alec's requirements above for completeness before I rewrite?

**Comment by jahilton** [comment]
LGTM.
Though I think instead of "There must be at least one valid embedding.", we should fail if any one embedding is invalid.

**Comment by atarashansky** [comment]
> LGTM. Though I think instead of "There must be at least one valid embedding.", we should fail if any one embedding is invalid.

This makes more sense to me - that way there's no confusion to users why explorer is missing a particular (invalid) embedding.

**Comment by brianraymor** [comment]
@atarashansky @dsadgat  @jahilton - could you take a look at the draft in the top-level summary comment? I've removed the screenshot of explorer for the moment that illustrated how {suffix} is surfaced in the UX. I will address that when I start editing #161 which will require some _restructuring_.

**Comment by brianraymor** [comment]
Are there additional constants that are prohibited/limited besides Infinity and NAN. For example, see: https://numpy.org/devdocs/reference/constants.html

**Comment by jahilton** [comment]
_one or more two-dimensional embeddings_
or
_one or more embeddings of at least two-dimensions_
?

**Comment by atarashansky** [comment]
@jahilton 
> one or more embeddings of at least two-dimensions
This one

**Comment by atarashansky** [comment]
> Are there additional constants that are prohibited/limited besides Infinity and NAN. For example, see: https://numpy.org/devdocs/reference/constants.html

Anything that yields a `True` for `np.isinf(DATA)` is invalid. We aren't checking for any other constants like negative zero.

**Comment by atarashansky** [comment]
Correct

**Comment by atarashansky** [comment]
The draft looks good to me as long as 
> MUST NOT contain any "Inf" values

is revised to something like
> MUST NOT contain any positive or negative infinity values

**Comment by brianraymor** [comment]
Yep - updated the draft above and included links to the numpy constants.

**Comment by dsadgat** [comment]
LGTM
